plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/TestBigQueryConnectorTest.java

```diff
@@ -331,6 +331,88 @@ protected boolean isColumnNameRejected(Exception exception, String columnName, b
     }
 
+    @Test
+    public void testPartitionDateColumn()
+    {
+        try (TestTable table = new TestTable(bigQuerySqlExecutor, "test.partition_date_column", "(value INT64) PARTITION BY _PARTITIONDATE")) {
+            // BigQuery doesn't allow omitting column list for ingestion-time partitioned table
+            // Using _PARTITIONTIME special column because _PARTITIONDATE is unsupported in INSERT statement
+            onBigQuery(format("INSERT INTO %s (_PARTITIONTIME, value) VALUES ('1960-01-01', 1)", table.getName()));
+            onBigQuery(format("INSERT INTO %s (_PARTITIONTIME, value) VALUES ('2159-12-31', 2)", table.getName()));
+
+            assertThat(query("SELECT value, \"$partition_date\" FROM " + table.getName()))
+                    .matches("VALUES (BIGINT '1', DATE '1960-01-01'), (BIGINT '2', DATE '2159-12-31')");
+
+            assertQuery(format("SELECT value FROM %s WHERE \"$partition_date\" = DATE '1960-01-01'", table.getName()), "VALUES 1");
+            assertQuery(format("SELECT value FROM %s WHERE \"$partition_date\" = DATE '2159-12-31'", table.getName()), "VALUES 2");
+
+            // Verify DESCRIBE result doesn't have hidden columns
+            assertThat(query("DESCRIBE " + table.getName())).projected(0).skippingTypesCheck().matches("VALUES 'value'");
+        }
+    }
+
+    @Test
+    public void testPartitionTimeColumn()
+    {
+        try (TestTable table = new TestTable(bigQuerySqlExecutor, "test.partition_time_column", "(value INT64) PARTITION BY DATE_TRUNC(_PARTITIONTIME, HOUR)")) {
+            // BigQuery doesn't allow omitting column list for ingestion-time partitioned table
+            onBigQuery(format("INSERT INTO %s (_PARTITIONTIME, value) VALUES ('1960-01-01 00:00:00', 1)", table.getName()));
+            onBigQuery(format("INSERT INTO %s (_PARTITIONTIME, value) VALUES ('2159-12-31 23:00:00', 2)", table.getName())); // Hour and minute must be zero
+
+            assertThat(query("SELECT value, \"$partition_time\" FROM " + table.getName()))
+                    .matches("VALUES (BIGINT '1', CAST('1960-01-01 00:00:00 UTC' AS TIMESTAMP(6) WITH TIME ZONE)), (BIGINT '2', CAST('2159-12-31 23:00:00 UTC' AS TIMESTAMP(6) WITH TIME ZONE))");
+
+            assertQuery(format("SELECT value FROM %s WHERE \"$partition_time\" = CAST('1960-01-01 00:00:00 UTC' AS TIMESTAMP(6) WITH TIME ZONE)", table.getName()), "VALUES 1");
+            assertQuery(format("SELECT value FROM %s WHERE \"$partition_time\" = CAST('2159-12-31 23:00:00 UTC' AS TIMESTAMP(6) WITH TIME ZONE)", table.getName()), "VALUES 2");
+
+            // Verify DESCRIBE result doesn't have hidden columns
+            assertThat(query("DESCRIBE " + table.getName())).projected(0).skippingTypesCheck().matches("VALUES 'value'");
+        }
+    }
+
+    @Test
+    public void testIngestionTimePartitionedTableInvalidValue()
```

Comment:
This test is here to verify BigQuery behaviour, correct?

I'm now confused about how can BigQuery have partitions `__NULL__` or `__UNPARTITIONED__` if it doesn't allow inserts with `NULL` or values outside the 1960 - 2159 range?

Reply:
Correct. A time partitioned table can have `__NULL__ ` and `__UNPARTITIONED__ ` partitions though ingestion time partitioned table disallow them in my understanding.